lib/spack/spack/cmd/develop.py

```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-            # are currently supported.
-            spec = spack.spec.parse_with_version_concrete(entry["spec"])
-            _retrieve_develop_source(spec, abspath)
 
-        if not env.dev_specs:
-            tty.warn("No develop specs to download")
+def setup_src_code(spec: spack.spec.Spec, src_path: str, clone: bool = True, force: bool = False):
+    """
+    Handle checking, cloning or overwriting source code
+    """
+    assert spec.versions
+    if clone is None:
+        clone = not os.path.exists(src_path)
 
-        return
+    if clone:
+        _clone(spec, src_path, force)
 
-    specs = spack.cmd.parse_specs(args.spec)
-    if len(specs) > 1:
-        raise SpackError("spack develop requires at most one named spec")
+    if not clone and not os.path.exists(src_path):
+        raise SpackError(f"Provided path {src_path} does not exist")
 
-    spec = specs[0]
 
-    version = spec.versions.concrete_range_as_version
-    if not version:
-        # look up the maximum version so infintiy versions are preferred for develop
-        version = max(spec.package_class.versions.keys())
-        tty.msg(f"Defaulting to highest version: {spec.name}@{version}")
-    spec.versions = spack.version.VersionList([version])
+def _update_config(spec, path):
+    find_fn = lambda section: spec.name in section
 
-    # If user does not specify --path, we choose to create a directory in the
-    # active environment's directory, named after the spec
-    path = args.path or spec.name
-    if not os.path.isabs(path):
-        abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
-    else:
-        abspath = path
+    entry = {"spec": str(spec)}
+    if path != spec.name:
+        entry["path"] = path
 
-    # clone default: only if the path doesn't exist
-    clone = args.clone
-    if clone is None:
-        clone = not os.path.exists(abspath)
+    def change_fn(section):
+        section[spec.name] = entry
 
-    if not clone and not os.path.exists(abspath):
-        raise SpackError("Provided path %s does not exist" % abspath)
+    spack.config.change_or_add("develop", find_fn, change_fn)
 
-    if clone:
-        if os.path.exists(abspath):
-            if args.force:
-                shutil.rmtree(abspath)
-            else:
-                msg = "Path %s already exists and cannot be cloned to." % abspath
-                msg += " Use `spack develop -f` to overwrite."
-                raise SpackError(msg)
 
-        _retrieve_develop_source(spec, abspath)
+def update_env_file(
+    env: spack.environment.Environment,
+    spec: spack.spec.Spec,
+    specified_path: Optional[str] = None,
+    build_dir: Optional[str] = None,
+):
+    """
+    Update the spack.yaml file with additions or changes from a develop call
+    """
+    tty.debug(f"Updating develop config for {env.name} transactionally")
+
+    # this check has to happen outside the write transaction otherwise updates don't seem to happen
```

Comment:
Can you fill me in a bit on what was happening with this? I don't understand why it would be necessary and that concerns me.

Reply:
This was just for unit-tests to pass. It worked from the CLI.  I think it had to do with process forks?  I'm not sure exactly. I spend a lot of time trying to debug but all my debugging instances were getting hung up in the subprocesses and the inspection/print statements looked like the values were correct, but the test always failed for updating the develop spec.

Reply:
Which test was the one failing? I'm hesitant to just ignore this since it suggests something wrong with how we handle environments more broadly.

Reply:
`test_develop_update_spec`

Reply:
With the current PR if I move this section back inside the `write_transaction` the following tests fail:

``` console
======================================================= short test summary info =======================================================
FAILED lib/spack/spack/test/cmd/develop.py::TestDevelop::test_develop_update_spec[mock_archive0] - AssertionError: assert 'mpich@=1.0' == 'mpich@=2.0'
FAILED lib/spack/spack/test/cmd/develop.py::test_recursive[mock_archive0] - AssertionError: assert 'indirect-mpich' in {'direct-mpich': {'path': 'direct-mpich', 'spec': 'direct-mpich@1.0/ldxba4og2st3nqkukqp...
```

Reply:
They pass if I check for the dev_entry outside of the `write_transaction`.